BUG: `MaskedNDArray.__array_ufunc__` breaks for `erfa.rxp`

I stumbled upon this while working on #16070. This looks like a bug in `MaskedNDArray.__array_ufunc__`, but maybe it's really a bug in pyerfa ?
I'm opening this report as soon as I managed to isolate the problem, but I suspect the issue is more general than it looks at first glance.

```python
import erfa
import numpy as np
from astropy.utils.masked import Masked
import pytest


@pytest.mark.parametrize(
    "m, v",
    [
        (np.eye(3), np.arange(3)), # this works
        (Masked(np.eye(3)), Masked(np.arange(3))), # this doesn't
    ],
)
def test_reg(m, v):
    erfa.rxp(m, v)
```

relevant pytest output
```
___________________________________________ test_reg[m1-v1] ___________________________________________

m = MaskedNDArray([[1., 0., 0.],
               [0., 1., 0.],
               [0., 0., 1.]])
v = MaskedNDArray([0, 1, 2])

    @pytest.mark.parametrize(
        "m, v",
        [
            (np.eye(3), np.arange(3)),
            (Masked(np.eye(3)), Masked(np.arange(3))),
        ],
    )
    def test_reg(m, v):
>       erfa.rxp(m, v)

g.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/clm/.pyenv/versions/astropy.dev/lib/python3.12/site-packages/erfa/core.py:19906: in rxp
    rp = ufunc.rxp(r, p)
../../astropy/utils/masked/core.py:872: in __array_ufunc__
    return self._masked_result(result, mask, out)
../../astropy/utils/masked/core.py:952: in _masked_result
    return Masked(result, mask)
../../astropy/utils/masked/core.py:77: in __new__
    return cls._get_masked_instance(*args, **kwargs)
../../astropy/utils/masked/core.py:125: in _get_masked_instance
    return masked_cls.from_unmasked(data, mask, copy)
../../astropy/utils/masked/core.py:518: in from_unmasked
    self._set_mask(mask, copy=copy)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MaskedNDArray([0., 1., 2.]), mask = array([[False],
       [False],
       [False]])
copy = False

    def _set_mask(self, mask, copy=False):
        self_dtype = getattr(self, "dtype", None)
        mask_dtype = (
            np.ma.make_mask_descr(self_dtype)
            if self_dtype and self_dtype.names
            else np.dtype("?")
        )
        ma = np.asanyarray(mask, dtype=mask_dtype)
        if ma.shape != self.shape:
            # This will fail (correctly) if not broadcastable.
            self._mask = np.empty(self.shape, dtype=mask_dtype)
>           self._mask[...] = ma
E           ValueError: could not broadcast input array from shape (3,1) into shape (3,)

../../astropy/utils/masked/core.py:238: ValueError
======================================= short test summary info =======================================
FAILED g.py::test_reg[m1-v1] - ValueError: could not broadcast input array from shape (3,1) into shape (3,)
```

--- a/astropy/utils/masked/core.py
+++ b/astropy/utils/masked/core.py
@@ -773,7 +773,7 @@ class MaskedNDArray(Masked, np.ndarray, base_cls=np.ndarray, data_cls=np.ndarray
                     ) = np.lib._function_base_impl._parse_gufunc_signature(
                         ufunc.signature.replace(" ", "")
                     )
-                axis = kwargs.get("axis", -1)
+                axis = kwargs.get("axis")
                 keepdims = kwargs.get("keepdims", False)
                 in_masks = []
                 for sig, mask in zip(in_sig, masks):
@@ -783,8 +783,13 @@ class MaskedNDArray(Masked, np.ndarray, base_cls=np.ndarray, data_cls=np.ndarray
                             # value in those is masked, the output will be
                             # masked too (TODO: for multiple core dimensions
                             # this may be too strong).
+                            in_axis = (
+                                tuple(range(-1, -1 - len(sig), -1))
+                                if axis is None
+                                else axis
+                            )
                             mask = np.logical_or.reduce(
-                                mask, axis=axis, keepdims=keepdims
+                                mask, axis=in_axis, keepdims=keepdims
                             )
                         in_masks.append(mask)
 
@@ -794,7 +799,10 @@ class MaskedNDArray(Masked, np.ndarray, base_cls=np.ndarray, data_cls=np.ndarray
                     if os:
                         # Output has core dimensions.  Assume all those
                         # get the same mask.
-                        result_mask = np.expand_dims(mask, axis)
+                        out_axis = (
+                            tuple(range(-1, -1 - len(os), -1)) if axis is None else axis
+                        )
+                        result_mask = np.expand_dims(mask, out_axis)
                     else:
                         result_mask = mask
                     result_masks.append(result_mask)

--- a/astropy/utils/masked/tests/test_functions.py
+++ b/astropy/utils/masked/tests/test_functions.py
@@ -173,6 +173,36 @@ class MaskedUfuncTests(MaskedArraySetup):
             assert res0.unmasked == exp[0]
             assert res0.mask == expected_mask[0]
 
+    def test_erfa_rxp(self):
+        # Regression tests for gh-16116
+        m = Masked(np.eye(3))
+        v = Masked(np.arange(6).reshape(2, 3))
+        rxp1 = erfa_ufunc.rxp(m, v)
+        exp = erfa_ufunc.rxp(m.unmasked, v.unmasked)
+        assert_array_equal(rxp1.unmasked, exp)
+        assert_array_equal(rxp1.mask, False)
+        v.mask[0, 0] = True
+        rxp2 = erfa_ufunc.rxp(m, v)
+        assert_array_equal(rxp2.unmasked, exp)
+        assert_array_equal(rxp2.mask, [[True] * 3, [False] * 3])
+        m.mask[1, 1] = True
+        v.mask[...] = False
+        rxp3 = erfa_ufunc.rxp(m, v)
+        assert_array_equal(rxp3.unmasked, exp)
+        assert_array_equal(rxp3.mask, True)
+
+    def test_erfa_rxr(self):
+        m1 = Masked(np.arange(27.0).reshape(3, 3, 3))
+        m2 = Masked(np.arange(-27.0, 0.0).reshape(3, 3, 3))
+        rxr1 = erfa_ufunc.rxr(m1, m2)
+        exp = erfa_ufunc.rxr(m1.unmasked, m2.unmasked)
+        assert_array_equal(rxr1.unmasked, exp)
+        assert_array_equal(rxr1.mask, False)
+        m1.mask[0, 1, 2] = True
+        rxr2 = erfa_ufunc.rxr(m1, m2)
+        assert_array_equal(rxr2.unmasked, exp)
+        assert np.all(rxr2.mask == [[[True]], [[False]], [[False]]])
+
     @pytest.mark.parametrize("axis", (0, 1, None))
     def test_add_reduce(self, axis):
         ma_reduce = np.add.reduce(self.ma, axis=axis)

--- a/docs/changes/utils/16120.bugfix.rst
+++ b/docs/changes/utils/16120.bugfix.rst
@@ -1,0 +1,2 @@
+Fix support in ``Masked`` for generalized ufuncs with more than a
+single core dimension (such as ``erfa.rxp``).

